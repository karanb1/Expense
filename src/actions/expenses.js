import uuid from 'uuid';
import database from '../firebase/firebase';
// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});
//dispatching an action fn using thunk
export const startAddExpense = (expenseData={})=>{
  return (dispatch) =>{
    const {
      description = '',
      note ='',
      amount= 0,
      createdAt = 0
    } = expenseData;
    const expense = {description,note,amount:amount/100,createdAt};
   return database.ref('expenses').push(expense).then((ref)=>{   //return only for asynchronous test case(promise chaining)
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
//Set expense
export const setExpense = (expenses)=>({
  type: "SET_EXPENSE",
  expenses
})
//dispatching an action fn using thunk
export const startSetExpense = () => {
  return (dispatch) => {
    const expenses = [];
    return database.ref(`expenses`).once('value')
      .then( (snapshot) => {
        snapshot.forEach( (childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          })
        })
        dispatch(setExpense(expenses));
      })
  }
}
  