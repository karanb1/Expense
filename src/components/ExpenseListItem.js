import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  
  <Link className="item" to={`/edit/${id}`}>
  <div className="listitem">
       <div>
           <h3>{description}</h3>
         <p className="time">{moment(createdAt).format('MMMM Do, YYYY')}</p>
      </div>
      <div>
           <h3>
           {numeral(amount).format('0,0.00')}
          </h3>
      </div>
      </div>
      </Link>
  
);

export default ExpenseListItem;
