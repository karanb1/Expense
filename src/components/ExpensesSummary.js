import React from 'react';
import { connect } from 'react-redux'
import expensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => (
    <div>
    {props.expenses[1] === 0 ? 
        (false)
        : 
        (props.expenses[1] === 1 ? 
        (<h1>Your total for {props.expenses[1]} expense is {numeral(props.expenses[0]).format('0,0.00')} INR </h1>)
       :
        (<h1>Your total for {props.expenses[1]} expenses is {numeral(props.expenses[0]).format('0,0.00')} INR </h1>))}
    </div>
)

const mapStateToProps =(state) => {
    return{
        expenses: expensesTotal(state.expenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);