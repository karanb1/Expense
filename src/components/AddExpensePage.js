import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component{
  onSubmit = (expense) =>{
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };

  render(){
    return(
      <div>
      <div className="summary" style={{paddingTop:40+'px', paddingBottom:40+'px'}}>
      <h1>Add Expense</h1>
      </div>
      <ExpenseForm
        onSubmit= {this.onSubmit}//after testing
      />
    </div>
    )
  }
}
const mapDispatchToProps = (dispatch)=> {
  return({
    startAddExpense: (expense)=> dispatch(startAddExpense(expense)) //can be many props so an object
  })
}


export default connect(undefined, mapDispatchToProps)(AddExpensePage);

//before testing
//     <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm
//       onSubmit= {(expense)=>{
//         props.dispatch(addExpense(expense));
//         props.history.push('/')
//       }}//after testing
//     />
//   </div>
//   )
// }
// }