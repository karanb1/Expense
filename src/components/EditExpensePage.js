import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component{
  onSubmit = (expense) => {
     this.props.onSubmit(this.props.expense.id,expense);
     this.props.history.push('/');
  };

  onRemove = ()=>{
    this.props.onClick({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render(){
    return(
      <div>
      <div className="summary" style={{paddingTop:40+'px', paddingBottom:40+'px'}}>
      <h1>Edit Expense</h1>
      </div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button className="remove" onClick={this.onRemove}>Remove</button>
    </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch,props)=>{
  return ({
    onSubmit: (id,expense) => dispatch(startEditExpense(id, expense)),
    onClick: (data)=> dispatch(startRemoveExpense(data))
    })
  }


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

// const EditExpensePage = (props) => {
//   return (
//     <div>
//       <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push('/');
//         }}
//       />
//       <button onClick={() => {
//         props.dispatch(removeExpense({ id: props.expense.id }));
//         props.history.push('/');
//       }}>Remove</button>
//     </div>
//   );
// };