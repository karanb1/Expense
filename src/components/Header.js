import React from 'react';
import { Link } from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth'

export const Header = (props) => (
  <header className="head">
    <Link className="title" to="/dashboard">
        <h1>Expense Tracker</h1>
    </Link>
    <button className="logout" onClick={props.startLogout}>Log out</button>
  </header>
);

const mapDispatchToProps = (dispatch)=>({
  startLogout: ()=> dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
