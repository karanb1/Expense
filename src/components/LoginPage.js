import React from 'react';
import {connect} from 'react-redux';
import {startLogin} from '../actions/auth'

export const LoginPage = (props) =>(
    <div className = "box-layout">
    <div className ="box">
         <h1>Expense Tracker</h1>
         <p>Start controlling your expenses</p>
        <button className= "button" onClick = {props.startLogin}>
        Login with  <i className="fab fa-google"> </i>
        </button>
    </div>
    </div>
);

const mapDispatchToProps = (dispatch)=>({
  startLogin: ()=> dispatch(startLogin())
})

export default connect(undefined,mapDispatchToProps)(LoginPage);