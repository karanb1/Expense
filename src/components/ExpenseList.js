import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div className="list">
   <div className="listtitle">
   <div className="mobile"><p>Expense</p></div>
   <div className="desktop"><p>Expense</p></div>
   <div className="desktop"><p>Amount</p></div>
   </div>
    {
      props.expenses.length === 0 ? (
        <p>No Expenses</p>
      ):(
      props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
      })
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
