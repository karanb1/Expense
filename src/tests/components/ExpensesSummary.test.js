import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses'
import expensesTotal from '../../selectors/expenses-total';

test('Should render expense total Correctly',()=>{
    const wrapper = shallow(<ExpensesSummary expenses={expensesTotal(expenses)} />)
    expect(wrapper).toMatchSnapshot();
})


