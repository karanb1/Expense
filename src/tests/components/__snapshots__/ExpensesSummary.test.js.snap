// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense total Correctly 1`] = `
<div>
  <h1>
    Your total for 
    3
     expenses is 
    1,141.95
     INR 
  </h1>
</div>
`;
